feat: default extra SANs for docker for a working cluster

faiq opened the pull request.

deepakm-ntnx committed:
[WIP] added nutanix support

dlipovetsky committed:
feat: Add user configuration for all providers

- Defines a cluster-level variable for defining one or more users
- Patches bootstrap templates for control plane and worker node pools
  with user configuration

faiq committed:
feat: adds nutanix csi to api

faiq committed:
feat: minor API tweaks and adds tooling for chart

faiq committed:
feat: more api changes

jimmidyson committed:
build: Remove unused example kustomizations

Missed when reorganising example kustomizations in previous PR.

faiq committed:
fix: adds proper permissions and api fixes

jimmidyson committed:
Merge pull request #1 from d2iq-labs/jimmi/remove-unused-kustomizations

faiq committed:
fix: adds configuration for new storage to examples/

faiq committed:
fix: don't use deepcopy

deepakm-ntnx committed:
added nutanix examples for clusterclass and cluster

deepakm-ntnx committed:
removed extra $ sign from all the variables

deepakm-ntnx committed:
Update hack/examples/sync.sh

Co-authored-by: Faiq <faiq.raza@nutanix.com>
Signed-off-by: Deepak Muley <deepak.muley@nutanix.com>

faiq committed:
fix: rename files to put the configmaps in the right place

deepakm-ntnx committed:
added new line to fix precommit lint

deepakm-ntnx committed:
fix: removed trailing whitespaces

faiq committed:
feat: use CAREN to set default cert SANS

deepakm-ntnx committed:
reverted devbox.lock changes

deepakm-ntnx committed:
fix: updated patches to use existing variable

faiq committed:
refactor: apply suggestion from review

faiq committed:
fix: linting errors

faiq committed:
feat: adds nutanix driver create storage class

faiq committed:
fix: pre-commit errors

jimmidyson committed:
test(e2e): Rename config for CAREN rather than CRE

jimmidyson committed:
Merge pull request #8 from d2iq-labs/jimmi/more-caren-renames

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Address review comments

faiq committed:
fix: pass in provisioner name

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Expand comments

faiq committed:
refactor: deploy storage class as a cluster resource set

jimmidyson committed:
test: Ensure defaults from JSON schema are respected

Without this, defaults declared in the JSON schema are not included in validation
steps, which can lead to invalid failures, while also not allowing for tests that
target defaults.

faiq committed:
refactor: change name of example sc

faiq committed:
fix: typo for default storage class key

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Fix typo in lockPassword logic, and add unit test

jimmidyson committed:
Merge pull request #9 from d2iq-labs/jimmi/variable-defaulting-in-tests

faiq committed:
fix: addon defaults for csi

faiq committed:
fix: sets the Items correctly

faiq committed:
fix: examples

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Add unit test for empty hashed password

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Change Sudo field from pointer to value

The zero value (empty string) is not valid, so the field does not need
to be a pointer.

faiq committed:
fix: actually apply cm

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Make username required

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Explain why we do not validate hashed password input

faiq committed:
fix: storage class crs

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Explain why we do not validate sudo input

faiq committed:
fix: use a unique storage class cm per cluster

faiq committed:
refactor: rename CSIProviders to CSI

faiq committed:
Merge pull request #5 from d2iq-labs/nutanix-examples

feat: Add examples for Nutanix provider

faiq committed:
fix: more unit tests

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Update type comments

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Remove errant comment

dlipovetsky committed:
fixup! feat: Add user configuration for all providers

Add users to the docs site

faiq committed:
fix: gomod

faiq committed:
test: adds a unit test for no params

dlipovetsky committed:
fixup! fixup! feat: Add user configuration for all providers

Manually wrap lines in doc

faiq committed:
feat: create crs from provider credentials

dkoshkin committed:
build: add Nutanix infra provider (#12)

Also deploy infra provider versions that match the API.

faiq committed:
feat: set a default storage class if 1 provider and 1 storage class config

faiq committed:
refactor: use default storage config obj

faiq committed:
feat: allow volume expansion

faiq committed:
feat: deploy snapshot chart with nutanix

dkoshkin committed:
fix: rename cpi to ccm (#11)

CPI is a term unique to the vSphere CCM.
Renaming to the more generic "CCM".

dlipovetsky committed:
test: Fix data race

dlipovetsky committed:
test: Check for data races in go tests

dkoshkin committed:
Merge pull request #15 from dlipovetsky/test-data-race

test: Fix data race

dlipovetsky committed:
test: Fix false negative in httpproxy test (#17)

Other patches can change the slice length, causing the previous check to
fail, although all correct patches are present.

dlipovetsky committed:
test: Fix false negative in auditpolicy test (#16)

Other patches can change the slice length, causing the previous check to
fail, although all correct patches are present.

faiq committed:
fix: set a new name for the tigera cm for that cluster (#14)

faiq committed:
fix: apply suggestions from reviews

dkoshkin committed:
fix: Nutanix examples (#19)

* fix: update latest template

* build: pin commit for nutanix examples

faiq committed:
Merge pull request #4 from dlipovetsky/users

feat: Add user configuration for all providers

faiq committed:
Merge pull request #7 from d2iq-labs/faiq/add-nutanix-csi

feat: add nutanix csi

faiq committed:
fix: removes unused cancel

faiq committed:
fix: adds nolint

faiq committed:
fix: adds missing items to scheme

dkoshkin commented on pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:
```diff
@@ -50,12 +61,16 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
+	if err := h.client.Get(ctx, clusterKey, cluster); err != nil {
+		return err
+	}
+	defaultAPICertSANs := getDefaultAPIServerSANs(cluster)
```
IIRC from the call last week, we wanted to also set defaults in the API either [here](https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/blob/7092a4c3ebfbe08d2f50a9a87209f4b5c14ea186/pkg/handlers/generic/clusterconfig/variables.go#L43) (if we get the required info of what infra provider it is) or [here](https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/blob/7092a4c3ebfbe08d2f50a9a87209f4b5c14ea186/api/v1alpha1/clusterconfig_types.go#L50), that way it shows up if the user doesn't set anything.
We also talked about adding the defaults in the handler as you have here so that the user doesn't need also worry about the defaults and break the cluster.

faiq replied:
That's not what I gathered. I thought this would be set no matter what because the wrong settings or not including these could break the cluster, which would need to be done in the handler.

dkoshkin replied:
I was thinking about this some more, I wonder if we are working around the limitation of not having a mutating webhook.

What we want is to show in the API what values are being set.
Setting it in openapi schema works for defaults, but then when a user overrides it we get these implicit defaults getting added in the handler then then doesn't show up in the `Cluster` object.

Should we finally work on adding the webhook? It would help with this work and other similar scenarios.

faiq replied:
Yes! this should definitely be a webhook, but I think that requires a bigger scope than this.

supershal replied:
> Setting it in openapi schema works for defaults, but then when a user overrides it we get these implicit defaults getting added in the handler then then doesn't show up in the Cluster object

Can you please explain a bit further? what needed to show up in `Cluster` object ? Wont extra SANs will be updated in `KubeadmControlPlaneTemplate` when the mutator adds them?

supershal replied:
Thanks for talking with me and explaining how webhook could work.

dkoshkin committed:
docs: use a KUBERNETES_VERSION env variable (#22)

* docs: use an env for KUBERNETES_VERSION

* docs: use a more widely available Kubernetes version

Makes the command more copy-paste friendly.
Kubernetes v1.28.7 is supported here by Docker, AWS and CAPX.

deepakm-ntnx committed:
feat: Add support for CAPX (#6)

* fix: added support for capx

* refactor: reuse existing CAPX types

* fix: set allowed enums for Nutanix resource types

* fix: set required for Nutanix node type

* fix: reuse resource.Quantity types

* fix: set defaults and validation

* fix: rename field to subnets

* refactor: fix handlers after API changes

* test: add new unit tests

* refactor: bring back host instead of address

* fix: examples with updated APIs

* fix: using latest capx private brach to test kube-vip fix

* fix: set namespace for credentialRef

The patch failed with the following error:
got failure response with message failed to apply JSON patches to input:
replace operation does not apply:
doc is missing key:
/spec/template/spec/prismCentral/credentialRef/namespace: missing value.

* docs: fix users example

* docs: deploying Calico for Nutanix

* fix: added basic docs for nutanix mutations

* fix: lint related fixes

* docs: minor changes

---------

Co-authored-by: Dimitri Koshkin <dimitri.koshkin@nutanix.com>

dkoshkin committed:
refactor: CCM code to support both CRS and CAAPH provider (#25)

supershal commented on api/v1alpha1/clusterconfig_types.go:
```diff
@@ -243,7 +251,10 @@ type ExtraAPIServerCertSANs []string
 	return clusterv1.VariableSchema{
 		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
-			Description: "Extra Subject Alternative Names for the API Server signing cert",
+			Description: fmt.Sprintf(
+				"Extra Subject Alternative Names for the API Server signing cert. For Docker %v are injected automatically.",
+				DefaultDockerCertSANs,
+			),
```
```suggestion
			Description: fmt.Sprintf(
				"Extra Subject Alternative Names for the API Server signing cert. For Docker %s are injected automatically.",
				strings.Join(DefaultDockerCertSANs,","),
			),
```

dkoshkin commented on pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:
```diff
@@ -50,12 +61,16 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
+	if err := h.client.Get(ctx, clusterKey, cluster); err != nil {
+		return err
+	}
+	defaultAPICertSANs := getDefaultAPIServerSANs(cluster)
```
WDYT leaving a TODO?
```suggestion
        TODO: move the defaults to a mutating webhook
	defaultAPICertSANs := getDefaultAPIServerSANs(cluster)
```

dkoshkin committed:
refactor: use a string type for Nutanix's AdditionalTrustBundle (#28)

* refactor: use a string type for Nutanix's AdditionalTrustBundle

Use a string instead of a ConfigMap reference
to make it easier for both the handlers to use, and the clients to set in the API.

* fix: force insecure: false with additionalTrustBundle

supershal commented:
WDYT of creating new separate patch `dockerExtraAPIServerCertSANsPatch` for Docker MetaVariable. Ensure that both lifecycle patch `extraAPIServerCertSANsPatch` and docker patch`dockerExtraAPIServerCertSANsPatch` preserves existing list of certSANs. 
i.e. something like 

for extraAPIServerCertSANsPatch
```
obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs = append(obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs, extraSANsFromVaraiable)
```

for dockerExtraAPIServerCertSANsPatch
```
obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs = append(obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs, dockerDefaultSANs)
```

supershal commented on pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:
```diff
@@ -64,11 +80,13 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
 		return nil
 	}
 
+	extraSans := deDup(extraAPIServerCertSANsVar, defaultAPICertSANs)
```
nit: AFIK dedup is not strictly needed. Duplicate SANs are allowed in Certs.

faiq committed:
build: update devbox.lock (#29)

supershal committed:
test: unit test for individual patch generator (#21)

* test: unit test for individual patch generator

* test: package level unit test for HTTPProxy

* test: fix data race between multiple unit test that use envtest

* test: make patchgenerator generic function

* fix: linting errors after rebase from main

supershal committed:
test: move all patch generator unit tests to their own packages (#23)

* test: unit test for individual patch generator

* test: package level unit test for HTTPProxy

* test: move region and httpproxy patch generator unit test invocation

* fix: linting errors

* test: move all AWS patch unit tests to their own packages (#24)

* test: move instanceprofile tests to its own package

* test: move instancetype unit tests to its own package

* test: move ami unit tests to its own package

* test: move aws network tests to its own package

* test: move controlplaneloadbalancer unit tests to its own package

* test: move aws cni unit tests to its own package

* test: fix linting errors

* test: unit tests for AWS security groups

* test: move customimage unit tests to their own package (#30)

* test:  move all Nutanix patch handler unit tests (#32)

* test: move controlplane endpoint unit tests

* test: move PC endpoint unit tests

* test: nove machinedetails unit tests

* test: move generic patch unit tests to own packages (#31)

* test: move audit policy tests to their own package

* test: move etcd unit tests to their own package

* test: move extra api server cert sans to its own package

* test: move image registry unit tests to its own package

* test: move kubernetes image repository unit tests

* test: move mirror unit tests

* test: move users unit tests

* test: remove gereric unit tests from nutanix meta patch handler

* test: cleaned up meta level unit test suites

dlipovetsky committed:
feat: Make containerd restart its own patch (#18)

* feat: Make containerd restart its own patch

* fix: unit tests for kubeadmconfigtemplate with containerdrestart patch

* fix: add comment for always add containerd patch

* test: move unit test to their own package

---------

Co-authored-by: Shalin Patel <shalin.patel@nutanix.com>

faiq committed:
fix: remove dedup as its not needed

supershal commented:
> WDYT of creating new separate patch `dockerExtraAPIServerCertSANsPatch` for Docker MetaVariable. Ensure that both lifecycle patch `extraAPIServerCertSANsPatch` and docker patch`dockerExtraAPIServerCertSANsPatch` preserves existing list of certSANs. i.e. something like

A bit more details: I think it would be cleaner to have provider specific logic in the provider's handler. i.e. `pkg/handlers/docker/mutation` and generic handlers can be free of any logic that is based on provider.  Having a docker mutator that always adds default SANs could be cleaner IMHO

faiq commented:
> >  WDYT of creating new separate patch dockerExtraAPIServerCertSANsPatch for Docker MetaVariable. Ensure that both lifecycle patch extraAPIServerCertSANsPatch and docker patchdockerExtraAPIServerCertSANsPatch preserves existing list of certSANs. i.e. something like

> A bit more details: I think it would be cleaner to have provider specific logic in the provider's handler. i.e. pkg/handlers/docker/mutation and generic handlers can be free of any logic that is based on provider. Having a docker mutator that always adds default SANs could be cleaner IMHO

I disagree. I think adding another handler that touches the same exact fields on objects can be more prone to errors and harder for a developer to discover what gets applied when
